infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/Bootstrapper.java

```diff
@@ -0,0 +1,176 @@
+
+import com.google.gson.Gson;
+import com.google.inject.assistedinject.Assisted;
+
+import org.eclipse.che.api.agent.shared.model.impl.AgentImpl;
+import org.eclipse.che.api.core.model.workspace.runtime.BootstrapperStatus;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.api.core.notification.EventService;
+import org.eclipse.che.api.core.notification.EventSubscriber;
+import org.eclipse.che.api.core.util.FileCleaner;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.shared.dto.RuntimeIdentityDto;
+import org.eclipse.che.api.workspace.shared.dto.event.BootstrapperStatusEvent;
+import org.eclipse.che.commons.lang.TarUtils;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Bootstraps agents installers.
+ *
+ * @author Sergii Leshchenko
+ */
+public class Bootstrapper {
+    private static final String BOOTSTRAPPER_BASE_DIR = "/home/user/che/bootstrapper/";
+    private static final String BOOTSTRAPPER_FILE     = "bootstrapper";
+    private static final String CONFIG_FILE           = "config.json";
+
+    private final    EventService                             eventService;
+    private final    String                                   machineName;
+    private final    RuntimeIdentity                          runtimeIdentity;
+    private final    DockerMachine                            dockerMachine;
+    private final    String                                   wsIntallerEndpoint;
+    private final    CountDownLatch                           bootstraplatch;
+    private final    AtomicInteger                            endpointId;
+    private final    EventSubscriber<BootstrapperStatusEvent> bootstrapperStatusListener;
+    private volatile BootstrapperStatusEvent                  resultEvent;
+
+    @Inject
+    public Bootstrapper(@Assisted String machineName,
+                        @Assisted RuntimeIdentity runtimeIdentity,
+                        @Assisted DockerMachine dockerMachine,
+                        EventService eventService,
+                        @Named("che.workspace.che_server_endpoint") String apiEndpoint) {
+        this.eventService = eventService;
+        this.machineName = machineName;
+        this.runtimeIdentity = runtimeIdentity;
+        this.dockerMachine = dockerMachine;
+        this.bootstraplatch = new CountDownLatch(1);
+        this.endpointId = new AtomicInteger();
+        this.wsIntallerEndpoint = apiEndpoint.replace("http", "ws").replace("wsmaster/api", "wsmaster")
```

Comment:
I think it makes sense to create a separate property for websocket endpoint, i think i saw similar code before.
Consider 
- separate property 
- dedicated [qualifier](https://github.com/google/guice/wiki/BindingAnnotations), so this property can be used like `@WebSocketEndpoint String wsEndpoint`